### PR TITLE
New multiple-variable entity constraints

### PR DESCRIPTION
**Pull request by bobular**
Resolves https://github.com/VEuPathDB/web-eda/issues/1410

There are two ways to test this PR

1. use your regular back end and this branch as-is. You will only be testing the front-end logic with spoofed back-end responses.
2. deploy a local back end using `DATA_TAG=branch-variable-constraints` and `RSERVE_TAG=3.2.1` and [use this branch from PR #1418 instead](https://github.com/VEuPathDB/web-eda/pull/1418) (which is a complete front and back end solution)

I'd appreciate it if you could cast your eyes over #1418 too.  I will merge that into this branch before merging this one.

Some info about the response spoofing, if you're using that:

There is a [function](https://github.com/VEuPathDB/web-eda/pull/1415/files#diff-b04aa83b90284c989ea432dda24f7697b0c7fe672c974242eb96a018b1d9127a) `temporaryHack()` which reconfigures the regular flat array from the back end into the `string[][]` shape we need, using various hard-coded hacks.  Before reporting any dependency-related issues, please check that the output of this function (logged in the console) is what was expected compared with changes made to the back end [here](https://github.com/VEuPathDB/EdaDataService/pull/216).

### CODE REVIEW

**Comment by d-callan** [comment]
wrt the 400, i looked into it briefly and it looks like its actually the merge service making that complaint. are you sure youve updated the requested entity/ output entity id in the post request?

File: src/lib/core/utils/data-element-constraints.ts [hunk 1]
```diff
@@ -319,3 +354,38 @@ export function mergeMaxNumValues(
...
+      entity.id === variable.entityId ||
+      entity.children?.includes(ancestors[0])
+    ) {
+      ancestors.unshift(entity);
+    }
+    return ancestors;
+  }, [] as StudyEntity[]);
+
+  return ancestors;
+}
```
Comment (dmfalke) [inline src/lib/core/utils/data-element-constraints.ts:376]: There is an assumption that `entities` are in a `preorder` traversal of the tree. That might be a risky assumption to make. At the very least, we should add a comment.

**Review by dmfalke — COMMENTED** [review]
I've looked over the code, and I have a couple of comments for now. I will read more thoroughly later, but I wanted to get these comments to you asap.

**Comment by bobular** [comment]
> wrt the 400, i looked into it briefly and it looks like its actually the merge service making that complaint. are you sure youve updated the requested entity/ output entity id in the post request?

Oh, no I haven't checked that. Will do. Thanks!

File: src/lib/core/hooks/findOutputEntity.ts [hunk 2]
```diff
@@ -23,7 +23,7 @@ export function useFindOutputEntity(
     if (providedEntityId)
       return entities.find((e) => e.id === providedEntityId);
     const variableName =
-      dataElementDependencyOrder?.[0] ?? fallbackVariableName;
+      dataElementDependencyOrder?.[0][0] ?? fallbackVariableName;
```
[the remark this answers is not in the document]
Comment (bobular) [reply src/lib/core/hooks/findOutputEntity.ts:26]: Aha, great. This hopfully addresses what Danielle said about the output entity.

Yes, they are both guaranteed to be on the "same branch" so no common ancestor required.

**Comment by dmfalke** [comment]
I'm going to test this out using a local backend. Thanks for the detailed instructions and very helpful set of PRs!

**Review by dmfalke — APPROVED** [review]
Works great!

**Comment by d-callan** [comment]
backend is merged

**Review by d-callan — APPROVED** [review]

**Comment by dmfalke** [comment]
@bobular I think you should be good to merge. I'll take care of deployment concerns to qa either tonight or tomorrow

**Pull request merged**